File: v4/packages/sdk-python/src/stagehand/locator.py
from __future__ import annotations

from collections.abc import Sequence
from typing import Literal, Self

from ._generated.models import (
    LocatorCentroidResult,
    LocatorClickOptions,
    LocatorClickParams,
    LocatorClickResult,
    LocatorCountResult,
    LocatorDescriptor,
    LocatorFillParams,
    LocatorFillResult,
    LocatorHighlightOptions,
    LocatorHighlightParams,
    LocatorHighlightResult,
    LocatorHoverResult,
    LocatorInnerHtmlResult,
    LocatorInnerTextResult,
    LocatorInputValueResult,
    LocatorIsCheckedResult,
    LocatorIsVisibleResult,
    LocatorScrollToParams,
    LocatorScrollToResult,
    LocatorSelectOptionParams,
    LocatorSelectOptionResult,
    LocatorSendClickEventOptions,
    LocatorSendClickEventParams,
    LocatorSendClickEventResult,
    LocatorTextContentResult,
    LocatorTypeOptions,
    LocatorTypeParams,
    LocatorTypeResult,
    MouseButton,
    RgbaColor,
)
from .rpc_client import RPCClient


class Locator:
    def __init__(
        self,
        rpc_client: RPCClient,
        *,
        page_id: str,
        selector: str,
        nth: int | None = None,
    ) -> None:
        self._rpc_client = rpc_client
        self._descriptor = LocatorDescriptor(page_id=page_id, selector=selector)
        if nth is not None:
            self._descriptor.nth = nth

    @property
    def page_id(self) -> str:
        return self._descriptor.page_id

    @property
    def selector(self) -> str:
        return self._descriptor.selector

    @property
    def nth_index(self) -> int | None:
        return self._descriptor.nth

    @property
    def descriptor(self) -> LocatorDescriptor:
        return self._descriptor

    async def click(
        self,
        *,
        button: MouseButton | Literal["left", "right", "middle"] | None = None,
        click_count: int | None = None,
    ) -> None:
        values = self._descriptor.model_dump(exclude_unset=True)
        options = LocatorClickOptions.model_validate({
            name: value
            for name, value in (("button", button), ("click_count", click_count))
            if value is not None
        })
        if options.model_fields_set:
            values["options"] = options
        await self._rpc_client.send(
            "locator.click",
            LocatorClickParams.model_validate(values),
            LocatorClickResult,
        )

    async def hover(self) -> None:
        await self._rpc_client.send(
            "locator.hover",
            self._descriptor,
            LocatorHoverResult,
        )

    async def fill(self, value: str) -> None:
        await self._rpc_client.send(
            "locator.fill",
            LocatorFillParams.model_validate({
                **self._descriptor.model_dump(exclude_unset=True),
                "value": value,
            }),
            LocatorFillResult,
        )

    async def count(self) -> int:
        result = await self._rpc_client.send(
            "locator.count",
            self._descriptor,
            LocatorCountResult,
        )
        return result.count

    async def is_checked(self) -> bool:
        result = await self._rpc_client.send(
            "locator.is_checked",
            self._descriptor,
            LocatorIsCheckedResult,
        )
        return result.checked

    async def input_value(self) -> str:
        result = await self._rpc_client.send(
            "locator.input_value",
            self._descriptor,
            LocatorInputValueResult,
        )
        return result.value

    async def is_visible(self) -> bool:
        result = await self._rpc_client.send(
            "locator.is_visible",
            self._descriptor,
            LocatorIsVisibleResult,
        )
        return result.visible

    async def inner_text(self) -> str:
        result = await self._rpc_client.send(
            "locator.inner_text",
            self._descriptor,
            LocatorInnerTextResult,
        )
        return result.text

    async def inner_html(self) -> str:
        result = await self._rpc_client.send(
            "locator.inner_html",
            self._descriptor,
            LocatorInnerHtmlResult,
        )
        return result.html

    async def text_content(self) -> str:
        result = await self._rpc_client.send(
            "locator.text_content",
            self._descriptor,
            LocatorTextContentResult,
        )
        return result.text_content

    async def scroll_to(self, percent: float | str) -> None:
        await self._rpc_client.send(
            "locator.scroll_to",
            LocatorScrollToParams.model_validate({
                **self._descriptor.model_dump(exclude_unset=True),
                "percent": percent,
            }),
            LocatorScrollToResult,
        )

    async def centroid(self) -> LocatorCentroidResult:
        return await self._rpc_client.send(
            "locator.centroid",
            self._descriptor,
            LocatorCentroidResult,
        )

    async def highlight(
        self,
        *,
        duration_ms: int | None = None,
        border_color: RgbaColor | None = None,
        content_color: RgbaColor | None = None,
    ) -> None:
        values = self._descriptor.model_dump(exclude_unset=True)
        options = LocatorHighlightOptions.model_validate({
            name: value
            for name, value in (
                ("duration_ms", duration_ms),
                ("border_color", border_color),
                ("content_color", content_color),
            )
            if value is not None
        })
        if options.model_fields_set:
            values["options"] = options
        await self._rpc_client.send(
            "locator.highlight",
            LocatorHighlightParams.model_validate(values),
            LocatorHighlightResult,
        )

    async def send_click_event(
        self,
        *,
        bubbles: bool | None = None,
        cancelable: bool | None = None,
        composed: bool | None = None,
        detail: float | None = None,
    ) -> None:
        values = self._descriptor.model_dump(exclude_unset=True)
        options = LocatorSendClickEventOptions.model_validate({
            name: value
            for name, value in (
                ("bubbles", bubbles),
                ("cancelable", cancelable),
                ("composed", composed),
                ("detail", detail),
            )
            if value is not None
        })
        if options.model_fields_set:
            values["options"] = options
        await self._rpc_client.send(
            "locator.send_click_event",
            LocatorSendClickEventParams.model_validate(values),
            LocatorSendClickEventResult,
        )

    async def type(self, text: str, *, delay: float | None = None) -> None:
        values = {**self._descriptor.model_dump(exclude_unset=True), "text": text}
        if delay is not None:
            values["options"] = LocatorTypeOptions(delay=delay)
        await self._rpc_client.send(
            "locator.type",
            LocatorTypeParams.model_validate(values),
            LocatorTypeResult,
        )

    async def select_option(self, values: str | Sequence[str]) -> list[str]:
        result = await self._rpc_client.send(
            "locator.select_option",
            LocatorSelectOptionParams.model_validate({
                **self._descriptor.model_dump(exclude_unset=True),
                "values": list(values) if not isinstance(values, str) else values,
            }),
            LocatorSelectOptionResult,
        )
        return result.values

    def first(self) -> Self:
        return self.nth(0)

    def nth(self, index: int) -> Self:
        return type(self)(
            self._rpc_client,
            page_id=self.page_id,
            selector=self.selector,
            nth=index,
        )

File: packages/sdk-ruby/lib/stagehand/locator.rb
# frozen_string_literal: true

require_relative "file_upload"
require_relative "generated/models"
require_relative "rpc_client"

module Stagehand
  # An element locator bound to a page, created via Page#locator. Port of
  # packages/sdk-python/src/stagehand/locator.py.
  class Locator
    attr_reader :page_id, :selector, :nth_index

    def initialize(rpc_client, page_id:, selector:, nth: nil)
      @rpc_client = rpc_client
      @page_id = page_id
      @selector = selector
      @nth_index = nth
    end

    def first
      nth(0)
    end

    def nth(index)
      self.class.new(@rpc_client, page_id: @page_id, selector: @selector, nth: index)
    end

    def click(button: nil, click_count: nil)
      values = descriptor_values
      options = { button: button, click_count: click_count }.compact
      values[:options] = Models::LocatorClickOptions.new(**options) unless options.empty?
      @rpc_client.send("locator.click", Models::LocatorClickParams.new(**values), "LocatorClickResult")
      nil
    end

    def fill(value)
      @rpc_client.send("locator.fill", Models::LocatorFillParams.new(**descriptor_values, value: value), "LocatorFillResult")
      nil
    end

    def hover
      @rpc_client.send("locator.hover", descriptor, "LocatorHoverResult")
      nil
    end

    # percent: 0-100 Numeric or a "50%" style String.
    def scroll_to(percent)
      @rpc_client.send("locator.scroll_to", Models::LocatorScrollToParams.new(**descriptor_values, percent: percent), "LocatorScrollToResult")
      nil
    end

    # The element's viewport centroid as Models::LocatorCentroidResult (x/y).
    def centroid
      @rpc_client.send("locator.centroid", descriptor, "LocatorCentroidResult")
    end

    # Colors are {r:, g:, b:, a:} Hashes or Models::RgbaColor.
    def highlight(duration_ms: nil, border_color: nil, content_color: nil)
      values = descriptor_values
      options = {
        duration_ms: duration_ms,
        border_color: rgba(border_color),
        content_color: rgba(content_color),
      }.compact
      values[:options] = Models::LocatorHighlightOptions.new(**options) unless options.empty?
      @rpc_client.send("locator.highlight", Models::LocatorHighlightParams.new(**values), "LocatorHighlightResult")
      nil
    end

    # Dispatches a synthetic MouseEvent("click") instead of trusted input.
    def send_click_event(bubbles: nil, cancelable: nil, composed: nil, detail: nil)
      values = descriptor_values
      options = { bubbles: bubbles, cancelable: cancelable, composed: composed, detail: detail }.compact
      values[:options] = Models::LocatorSendClickEventOptions.new(**options) unless options.empty?
      @rpc_client.send("locator.send_click_event", Models::LocatorSendClickEventParams.new(**values), "LocatorSendClickEventResult")
      nil
    end

    # values: a String or array of Strings; returns the selected values.
    def select_option(values)
      @rpc_client.send(
        "locator.select_option",
        Models::LocatorSelectOptionParams.new(**descriptor_values, values: values.is_a?(String) ? values : values.to_a),
        "LocatorSelectOptionResult",
      )
    end

    def type(text, delay: nil)
      values = descriptor_values
      values[:text] = text
      values[:options] = Models::LocatorTypeOptions.new(delay: delay) unless delay.nil?
      @rpc_client.send("locator.type", Models::LocatorTypeParams.new(**values), "LocatorTypeResult")
      nil
    end

    def count
      @rpc_client.send("locator.count", descriptor, nil)
    end

    def text_content
      @rpc_client.send("locator.text_content", descriptor, nil)
    end

    def inner_text
      @rpc_client.send("locator.inner_text", descriptor, nil)
    end

    def inner_html
      @rpc_client.send("locator.inner_html", descriptor, nil)
    end

    def input_value
      @rpc_client.send("locator.input_value", descriptor, nil)
    end

    def visible?
      @rpc_client.send("locator.is_visible", descriptor, nil)
    end
    alias is_visible visible?

    def checked?
      @rpc_client.send("locator.is_checked", descriptor, nil)
    end
    alias is_checked checked?

    def set_input_files(files)
      values = descriptor_values
      values[:files] = FileUpload.normalize_file_input(files)
      @rpc_client.send("locator.set_input_files", Models::LocatorSetInputFilesParams.new(**values), "LocatorSetInputFilesResult")
      nil
    end

    private

    def rgba(color)
      case color
      when nil, Models::RgbaColor then color
      when Hash then Models::RgbaColor.new(**color.transform_keys(&:to_sym))
      else raise ArgumentError, "colors must be a Models::RgbaColor or a Hash with r/g/b (and optional a)"
      end
    end

    def descriptor
      Models::LocatorDescriptor.new(**descriptor_values)
    end

    def descriptor_values
      values = { page_id: @page_id, selector: @selector }
      values[:nth] = @nth_index unless @nth_index.nil?
      values
    end
  end
end
